pages/blog/internals-of-nova-part-2.md

```diff
@@ -0,0 +1,286 @@
+
+Data-oriented design is all about solving or at least mitigating this issue. To do so, we must know our data and use that knowledge to minimize memory usage which always improves cache usage, and we must take into account the CPU caches when desigining our software. The important thing here is that a CPU loads data into cache not byte-by-byte but in cache lines. This is usually 64 bytes at a time, though recent CPUs have bumped this number up to 128 bytes.
+
+```rs
+#[repr(align(64))]
+struct CacheLine([u8; 64]);
+```
+
+When we load a piece of data we load a full cache line into the CPU, no matter how small a part of it we actually want to use. Loading this one cache line from main memory takes 200 nanoseconds. Now, if we only use a single byte (or worse, one bit!) of that data then the vast majority of the data loaded is wasted. What we want is to ensure that as much of the cache line gets used, especially if we are in a tight loop.
+
+Let's take as an example an Array of Arrays in the V8 JavaScript engine.
+
+```ts
+const arr: Data[][] = await Promise.all(names.map(fetchDataForName));
+```
+
+Now say we are interested in how many `Data` entities there are in total: We will loop over the `arr` Array, accessing each Array in it, and tallying up the length:
+
+```ts
+const dataCount = arr.reduce((acc, dataArray) => acc + dataArray.length, 0);
+```
+
+In terms of memory access patterns, a V8 Array looks roughly like this:
+
+```rs
+struct Array(*mut ArrayHeapData);
+
+// Size is 32 bytes in Node, 16 bytes in Chrome with pointer compression
+struct ArrayHeapData {
+    map: *mut Map,
+    elements: *mut FixedArray,
+    length: u32,
+    properties: *mut FixedArray,
+}
+```
+
+When we reduce through `arr`, we must first access the `elemeents`. In this case it gives us roughly something like:
+
+```rs
+struct FixedArray<const N: usize> {
+    length: usize,
+    entries: [*mut ArrayHeapData; N],
+}
+```
+
+From here we can chase each `*mut ArrayHeapData` to get to the `length` data contained in each, which we then load and sum up.
+
+It is fair to assume that most of these `ArrayHeapData` structs are located close to each other: We shall assume that they are actually right next to one another which means that we can fit two of them on one cache line (four in Chrome). What we want out of each `ArrayHeapData` is a single `u32`, that is 4 bytes. This means that we use 8 bytes (16 in Chrome) of every 64-byte cache line. That is a usage rate of 12.5%. In expert circles this is called "bad".
```

Comment:
**confusion**: Why is it double in chrome?

Reply:
An Array's data in V8 without pointer compression takes up 4 * 8 bytes, of which the `length` part is either an SMI (31 bit number) or a pointer to a HeapNumer if the length is over the SMI limit. Effectively then, it's ~always a ~32 bit integer.

With pointer compression, which is in use in Chrome/Chromium but not in Node, all pointers become just 4-byte indexes from the beginning of the heap (so 0 is beginning of heap, 1 is 4 bytes from the beginning, ...) instead. So now an Array's data takes up 4 * 4 bytes and the `length` part is again either an SMI or an index to a HeapNumber. But it's still effectively always a ~32 bit number, so we're still reading the same amount of data, but more of these fit on a single cache line (4 when it was 2 before). Thus, the amount of the cache line that is actually interesting to us doubles.